### ISSUE TITLE
Try to implement FCN but get some errors

### ISSUE THREAD

**Issue by tahaemara** [deeplearning4j/deeplearning4j#6043]
I had tried to implement a semantic segmentation architecture FCN-32s, but I faced an error.

`Exception in thread "main" java.lang.RuntimeException: Failed to allocate 20480864256 bytes from DEVICE [0] memory
	at org.nd4j.jita.memory.CudaMemoryManager.allocate(CudaMemoryManager.java:60)
	at org.nd4j.jita.workspace.CudaWorkspace.alloc(CudaWorkspace.java:198)
	at org.nd4j.jita.allocator.impl.AtomicAllocator.allocateMemory(AtomicAllocator.java:454)
	at org.nd4j.jita.allocator.impl.AtomicAllocator.allocateMemory(AtomicAllocator.java:399)
	at org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer.<init>(BaseCudaDataBuffer.java:269)
	at org.nd4j.linalg.jcublas.buffer.CudaFloatDataBuffer.<init>(CudaFloatDataBuffer.java:64)
	at org.nd4j.linalg.jcublas.buffer.factory.CudaDataBufferFactory.createFloat(CudaDataBufferFactory.java:323)
	at org.nd4j.linalg.factory.Nd4j.createBuffer(Nd4j.java:1481)
	at org.nd4j.linalg.factory.Nd4j.createBuffer(Nd4j.java:1455)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.<init>(BaseNDArray.java:248)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.<init>(BaseNDArray.java:285)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.<init>(BaseNDArray.java:626)
	at org.nd4j.linalg.jcublas.JCublasNDArray.<init>(JCublasNDArray.java:257)
	at org.nd4j.linalg.jcublas.JCublasNDArrayFactory.create(JCublasNDArrayFactory.java:257)
	at org.nd4j.linalg.factory.Nd4j.create(Nd4j.java:5085)
	at org.nd4j.linalg.workspace.BaseWorkspaceMgr.create(BaseWorkspaceMgr.java:240)
	at org.deeplearning4j.nn.layers.convolution.Deconvolution2DLayer.preOutput(Deconvolution2DLayer.java:198)
	at org.deeplearning4j.nn.layers.convolution.Deconvolution2DLayer.activate(Deconvolution2DLayer.java:233)
	at org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doForward(LayerVertex.java:105)
	at org.deeplearning4j.nn.graph.ComputationGraph.ffToLayerActivationsInWS(ComputationGraph.java:1946)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1310)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1280)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:178)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:60)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:54)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1104)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1036)
	at com.emaraic.semanticsegmentation.FCN.main(FCN.java:140)
	Suppressed: java.lang.RuntimeException: Failed to allocate 21879030515 bytes from HOST memory
		at org.nd4j.jita.memory.CudaMemoryManager.allocate(CudaMemoryManager.java:47)
		at org.nd4j.jita.workspace.CudaWorkspace.init(CudaWorkspace.java:63)
		at org.nd4j.linalg.memory.abstracts.Nd4jWorkspace.initializeWorkspace(Nd4jWorkspace.java:486)
		at org.nd4j.linalg.memory.abstracts.Nd4jWorkspace.close(Nd4jWorkspace.java:627)
		at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1355)
		... 7 more
`

the class RecordLoader.java is used for loading input image(3channel) and also label image (1 channel each pixel represents the class )  as the input to computationalgraph is 3c image and the output is 19(num of classes) matrices (each 1 channel), I think that there is something I missed in class  [RecordLoader.java](https://gist.github.com/tahaemara/411c715160598530ea3b97b60661b121#file-recordloader-java-L89) and also in the final layes in computation graph

I followed [ this code](https://github.com/apache/incubator-mxnet/blob/master/example/fcn-xs/symbol_fcnxs.py) without using any pre-trained model, it the next step I will do after solving this issue.

I attached the complete error output and the  source code to this gist
https://gist.github.com/tahaemara/411c715160598530ea3b97b60661b121

Thanks in advance,

**Comment by raver119** [comment]
Sorry, but what's actual question? Your exception is just OOM - model tried to allocate 20GB of memory.
Not too much options here:
- decrease batch size
- tweak architecture
- stick to CPU

p.s. make sure you're using cuDNN. It'll somewhat reduce memory requirements.

**Comment by saudet** [comment]
If you're using cuCNN, disable it's workspace allocations: https://deeplearning4j.org/cudnn
